internal/future/statestore/redis.go

```diff
@@ -94,29 +96,208 @@ func redisCloser(closer func() error) {
+// DeleteTicket removes the Ticket with the specified id from state storage.
+func (rb *redisBackend) DeleteTicket(ctx context.Context, id string) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"error": err.Error(),
+		}).Error("failed to connect to redis")
+		return err
+	}
+	defer redisConn.Close()
+
+	_, err = redisConn.Do("DEL", id)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"key":   id,
+			"cmd":   "DEL",
+			"error": err.Error(),
+		}).Error("failed to delete the ticket from state storage")
+		return err
+	}
+
+	return nil
 }
 
 // IndexTicket indexes the Ticket id for the specified fields.
-func (rb *redisBackend) IndexTicket(context.Context, pb.Ticket, []string) error {
-	return status.Error(codes.Unimplemented, "not implemented")
+func (rb *redisBackend) IndexTicket(ctx context.Context, ticket pb.Ticket, indices []string) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"error": err.Error(),
+		}).Error("failed to connect to redis")
+		return err
+	}
+	defer redisConn.Close()
+
+	redisConn.Send("MULTI")
+	// Loop through all attributes we want to index.
+	for _, attribute := range indices {
+		// TODO: Currently, Open Match supports specifying attributes as JSON properties on Tickets.
+		// Alternatives for populating this information in proto fields are being considered.
+		// Also, we need to add Ticket creation time to either the ticket or index.
+		// Meta characters bein specified in JSON property keys is currently not supported.
+		v := gjson.Get(ticket.Properties, attribute)
+
+		// If this attribute wasn't provided in the JSON, continue to the next attribute to index.
+		if !v.Exists() {
+			redisLogger.WithFields(logrus.Fields{
+				"attribute": attribute}).Warningf("Couldn't find index in Ticket Properties")
+			continue
+		}
+
+		// Value exists. Check if it is a supported value for indexed properties.
+		if v.Int() < -9223372036854775808 || v.Int() > 9223372036854775807 {
+			redisLogger.WithFields(logrus.Fields{
+				"attribute": attribute}).Warningf("Invalid value for attribute, skip indexing.")
+			continue
+		}
+
+		// Index the attribute by value.
+		value := v.Int()
+		redisConn.Send("ZADD", attribute, value, ticket.Id)
+	}
+
+	// Run pipelined Redis commands.
+	_, err = redisConn.Do("EXEC")
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"id":    ticket.Id,
+			"error": err.Error(),
+		}).Error("failed to index the ticket")
+		return err
+	}
+
+	return nil
 }
 
 // DeindexTicket removes the indexing for the specified Ticket. Only the indexes are removed but the Ticket continues to exist.
-func (rb *redisBackend) DeindexTicket(context.Context, string, []string) error {
-	return status.Error(codes.Unimplemented, "not implemented")
+func (rb *redisBackend) DeindexTicket(ctx context.Context, id string, indices []string) error {
```

Comment:
Do we need to check the existence of the passed-in ticket?

Reply:
Deindexing will be successful - there may possibly be indexes that were leaked for a ticket and ticket actually deleted. I think its ok to delete (so long as non existing are treated as success) than add an additional existence check.